added base profiler

kurisusnowdeng opened the pull request.

FrankLeeeee commented on colossalai/utils/profiler/common.py:
```diff
@@ -0,0 +1,75 @@
+from torch.utils.tensorboard import SummaryWriter
+
+
+class ProfilerEvent(Enum):
+    '''
+    Event types to profile
+    '''
+    COMM = CommProfiler
```
why is profiler an event?

kurisusnowdeng replied:
It is to distinguish events (or activities? as named by torch) to profile (e.g. communication, memory), and each event relates to a profiler

FrankLeeeee replied:
It is kind of confusing as in communication profiler, it has defined its own event.

kurisusnowdeng replied:
So change it to 'ProfilerActivity'?

FrankLeeeee replied:
ok

FrankLeeeee commented on colossalai/utils/profiler/common.py:
```diff
@@ -0,0 +1,75 @@
+from enum import Enum
+from .comm_profiler import CommProfiler
+from typing import List
+from torch.utils.tensorboard import SummaryWriter
```
should catch import error here as this requires tensorboard installation

kurisusnowdeng replied:
Pytorch already helps raise the import error

FrankLeeeee replied:
This line makes TensorBoard a must.

kurisusnowdeng replied:
Do you mean even there is no tensorboard installed we still make profiler work?

FrankLeeeee replied:
Yes

FrankLeeeee replied:
In fact, a better workaround of this is to import tensorboard only in `to_tensorboard`. Type hinting can be given as a string e.g. 

```python
writer: "torch.utils.tensorboard.SummaryWriter"
```

FrankLeeeee commented on colossalai/utils/profiler/_base_profiler.py:
```diff
@@ -0,0 +1,28 @@
+class BaseProfiler(object):
+
+    def __init__(self):
+        ...
```
this `...` will raise error

kurisusnowdeng replied:
Which error? I tried a simple test, and it was fine.

FrankLeeeee replied:
Not an error, but this function should be removed if no initialization is specified.

FrankLeeeee commented on colossalai/utils/profiler/_base_profiler.py:
```diff
@@ -0,0 +1,28 @@
+from torch.utils.tensorboard import SummaryWriter
+
+
+class BaseProfiler(object):
```
make it an abstract class with abc

feifeibear replied:
derive from ABC if it's an abstract class

FrankLeeeee commented on colossalai/utils/profiler/common.py:
```diff
@@ -0,0 +1,75 @@
+        for e in self.events:
+            e.to_tensorboard(writer)
+
+    def to_file(self, filename: str):
+        for e in self.events:
+            e.to_file(filename)
```
multiple profilers write into the same file?

kurisusnowdeng replied:
How about provide a path and add 'profilerevent_name.log' as a suffix

FrankLeeeee replied:
One way to handle this is to set the format as {output_path}/{prefix}_{profiler.__class__.__name__}_{timestamp}.log. This method requires changing the parameter filename to otuput_path

FrankLeeeee commented on colossalai/utils/profiler/common.py:
```diff
@@ -0,0 +1,75 @@
+        return res
+
+
+def colossalai_profiler(*args, **kwargs):
+    '''
+    Profiler context manager
+
+    Usage:
+    from colossalai.utils.profiler import ProfilerEvent, colossalai_profiler
+    from torch.utils.tensorboard import SummaryWriter
+
+    with colossalai_profiler(events=[ProfilerEvent.COMM]) as prof:
+        train()
+    writer = SummaryWriter('tb/path')
+    prof.to_tensorboard(writer)
+    prof.to_file('./prof.log')
+    prof.show()
+    '''
+    return ProfilerContext(*args, **kwargs)
```
This wrapper does not do anything, can just let user import ProfilerContext

kurisusnowdeng replied:
Just mean to use lowercase api for context

FrankLeeeee replied:
It is alright to use `ProfilerContext`. In fact, the `*args` and `**kwargs` can hurt usability.

FrankLeeeee commented on colossalai/utils/profiler/common.py:
```diff
@@ -0,0 +1,75 @@
+    '''
+    Event types to profile
+    '''
+    COMM = CommProfiler
```
Do not use abbreviation for API exposed to the user.

FrankLeeeee replied:
It is not a common practice to put a class as the enum value.  It seems not standard to me even though it works. What do you think @feifeibear .

feifeibear replied:
Also weird to me

FrankLeeeee replied:
@kurisusnowdeng Any idea about this part?

kurisusnowdeng replied:
We can also use some keys to redirect to the profiler classes in a dict. But, I think this is a redundant step.

xcetos replied:
Actually, using Enum as an option is quite common in C++ and Rust, but I have no idea whether this is appropriate for Python.

FrankLeeeee commented on colossalai/utils/profiler/common.py:
```diff
@@ -0,0 +1,87 @@
+    COMM = CommProfiler
+
+
+class colossalai_profiler(object):
```
Follow python class naming convention

FrankLeeeee commented on colossalai/utils/profiler/common.py:
```diff
@@ -0,0 +1,87 @@
+    COMM = CommProfiler
+
+
+class colossalai_profiler(object):
```
Follow python class naming convention

FrankLeeeee replied:
we can directly do this

```python
with Profiler(...) as prof:
    do_something()
```

I don't think we need to have the repository name in the class name.

kurisusnowdeng replied:
Sure, but conventionally context classes are lowercase

FrankLeeeee replied:
I see, I can accept the current code then.

kurisusnowdeng commented:
@FrankLeeeee All fixed.

FrankLeeeee commented on colossalai/utils/profiler/__init__.py:
```diff
@@ -1 +1,2 @@
-from .comm_profiler import enable_communication_prof, communication_prof_show
+# from .comm_profiler import enable_communication_prof, communication_prof_show
+from .common import ProfilerActivity, colossalai_profiler
```
Will this change cause the test_profiler unit testing to fail?

kurisusnowdeng replied:
Definitely. Need to update unitests after the update of specific profilers.

kurisusnowdeng replied:
Sure. Skip it.

kurisusnowdeng commented:
@FrankLeeeee Fixed.

FrankLeeeee commented:
Unit testing failed.

kurisusnowdeng commented:
Updated. Should be fine now. Also changed the way to initialize the profiler context into passing profiler instances in, according to the suggestion from @1SAA and @xcetos 。

FrankLeeeee commented on colossalai/utils/profiler/common.py:
```diff
@@ -0,0 +1,83 @@
+    from colossalai.utils.profiler import ProfilerActivity, profiler
+    from torch.utils.tensorboard import SummaryWriter
+
+    with profiler(activities=[ProfilerActivity.COMMUNICATION]) as prof:
+        train()
```
This should be updated.

FrankLeeeee commented on colossalai/utils/profiler/common.py:
```diff
@@ -0,0 +1,83 @@
+    prof.show()
+    '''
+
+    def __init__(self, activities: List[BaseProfiler] = None):
```
shall we change the parameter name as the type is a profiler

kurisusnowdeng replied:
I think 'activities' also makes sense, and maybe easier for pytorch users to distinguish the specific profilers from the profiler context (just too many 'profiler's).

kurisusnowdeng replied:
@FrankLeeeee BTW, ready to merge if there is no other problems.

FrankLeeeee commented:
According to @1SAA , this PR is replaced by #340 so I will close this PR now.
